### PR TITLE
[Feature] Multicollector interruptor

### PR DESCRIPTION
**Pull request by albertbou92**
## Description

This PR implements a preemptive mechanism to early stop stragglers in multi sync data collectors. The invalid data can be identified because its trajectory ids are -1.

For now, preemptive mechanism is not compatible with `split_trajs=True`, but this could be easily adapted by ignoring trajectory ids equal to -1 in the `split_trajectories` method.

i.e. 
splits = [(splits == i).sum().item() for i in splits.unique_consecutive() if i != -1] (changing line 39 in collectors/utils.py)
out_splits = rollout_tensordict.view(-1)[traj_ids != -1].split(splits, 0) (changing line 54 in collectors/utils.py)

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close #15213` if this solves the issue #15213

- [ ] I have raised an issue to propose this change ([required](https://github.com/pytorch/rl/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: torchrl/collectors/collectors.py [hunk 1]
```diff
@@ -67,6 +68,44 @@ def __call__(self, td: TensorDictBase) -> TensorDictBase:
         return td.set("action", self.action_spec.rand())
 
 
+class Interruptor:
```
Comment (vmoens) [inline torchrl/collectors/collectors.py:71]: if it's public it should be in the doc

File: torchrl/collectors/collectors.py [hunk 2]
```diff
@@ -67,6 +68,44 @@ def __call__(self, td: TensorDictBase) -> TensorDictBase:
...
+    def stop_collection(self):
+        with self._lock:
+            self._collect = False
+
+    def collection_stopped(self):
+        with self._lock:
+            return self._collect is False
+
+
+class InterruptorManager(SyncManager):
```
Comment (vmoens) [inline torchrl/collectors/collectors.py:96]: ditto

File: test/test_collector.py [hunk 3]
```diff
@@ -1186,6 +1187,72 @@ def weight_reset(m):
...
+            reset_at_each_iter=False,
+            devices="cpu",
+            storing_devices="cpu",
+            split_trajs=False,
+            preemptive_threshold=0.25,
+        )
+
+        for batch in collector:
+            assert (
+                batch["collector"]["traj_ids"][
```
Comment (vmoens) [inline test/test_collector.py:1249]: let's rewrite this in a more straightforward way, it's a bit hard to read
What if none is -1? Shouldn't we test that we have at least one traj_id set to -1 in the batch?

**Review by vmoens — COMMENTED** [review]
LGTM I left some comments

File: test/test_collector.py (hunk 3 above)
Comment (albertbou92) [reply test/test_collector.py]: I simplified the code for more clarity.

Regarding the possibility of having none -1, I have addressed that as well by setting the preemptive threshold to 0.0. This way, all collectors will stop after the first iteration and only the very first frame of each sync collector will be valid. So we know for sure there are -1's in the batch.

**Comment by vmoens** [comment]
Can you merge main in this branch, the tests are failing bc they're looking for a deprecated function in tensordict

**Review by vmoens — COMMENTED** [review]
Can you have a look at how it would play out with the distributed collectors I integrated yesterday?
We don't necessarily need everything to be fully compatible but I want to make sure that we're not missing some obvious point, e.g. how we handle the traj-ids in the distributed collector may need a bit of refactoring.

(I will merge this sooner than that as the solution is neat and usable as of now)

File: torchrl/collectors/collectors.py [hunk 4]
```diff
@@ -1399,6 +1462,18 @@ def iterator(self) -> Iterator[TensorDictBase]:
 
             i += 1
             max_traj_idx = None
+
+            if self.interruptor:
```
Comment (vmoens) [inline torchrl/collectors/collectors.py:1466]: for clarity can we have `if self.interruptor is not None`?

**Pull request merged**